Integrate protocol interface

filmaj opened the pull request:
###  Summary

This PR introduces use of the various CLI/SDK protocol implementations provided by https://github.com/slackapi/deno-slack-protocols.

Also fixes up broken code coverage reporting to codecov since the introduction of the prettier CLI reporter.

filmaj commented on src/local-run.ts:
```diff
@@ -72,6 +78,13 @@ export const getCommandline = function (
+  // add them here.
+  if (hookCLI.getCLIFlags) {
+    const flags = hookCLI.getCLIFlags();
+    command.push(...flags);
```
This would only be relevant to the MessageBoundaryProtocol.

That protocol relies on additional CLI flags being provided by the CLI (in particular, `--boundary=somestring`, which would include a random-generated string created by the CLI on each hook invocation).

The way the local-run `start` hook works in this repo is:

- CLI invokes the `local-run.ts` method as a new process...
- `local-run.ts` in turn invokes _another_ process (`local-run-function.ts`) <-- we need to make sure any protocol-specific flags get passed to the child process here. That's what this code does.

WilliamBergamin replied:
Could something like a `CLIContext` object be helpful here or will this type of information always be passed through flags?

filmaj replied:
Maybe one day if the protocol negotiation changes, but as of right now, we will rely on flags. The only other process-specific interfaces we could leverage to pass context around that I can think of would include environment variables, flags to the command, or `stdin`. If we ever do reach for those as part of protocol negotiation / hook invocation, we can then encapsulate that in a new abstraction as you suggest, but, until then, I don't think we need it.

shapirone commented on src/local-run.ts:
```diff
@@ -101,18 +113,18 @@ export const runWithOutgoingDomains = async function (
   } catch (e) {
-    log("Error determining deno executable path: ", e);
+    hookCLI.error("Error determining deno executable path: ", e);
+    // TODO: should we throw here?
```
IIRC we were intentionally not throwing here

shapirone reviewed (approved):
This LGTM, added @curtisallen as a reviewer. I didn't see any issues in local run or on ROSI

filmaj commented:
Good call on pinging @curtisallen, that reminds me, I should deploy this runtime version out to slackdev and put it through the gears... will do that today before flying out.

curtisallen reviewed (approved):
This LGTM. Thanks!

filmaj commented:
I added a couple of tests and rewrote some of the tests a bit to adhere to our new test writing style, as well as simplify some stubs with type assertions. Also updated the protocol library to use the latest released version. @shapirone can you do one more once-over?

shapirone commented on src/deps.ts:
```diff
@@ -1,4 +1,10 @@
 export { readAll } from "https://deno.land/std@0.99.0/io/util.ts";
 export { BaseSlackAPIClient } from "https://deno.land/x/deno_slack_api@0.0.2/base-client.ts";
+// TODO: can keep this dependency on deno_slack_builder until the next release, then switch this over to pull this from deno-slack-hooks.
+//       otherwise we'll have to be careful on the release timing of this repo and the hooks repo (have to both appear at the same time)
+//       luckily, there are no concerns about different protocols impacting the use of createManifest imported here, as the below is only
+//       used during local-run, which in the default/legacy protocol, logging to stdout is unlimited for that specific hook.
```
nit: Feels like the content of this TODO is less applicable after we merge

```suggestion
// TODO: Update to import from `deno-slack-hooks` instead
```

shapirone commented on src/dev_deps.ts:
```diff
@@ -8,4 +8,6 @@ export {
   fail,
 } from "https://deno.land/std@0.152.0/testing/asserts.ts";
 export * as mock from "https://deno.land/std@0.152.0/testing/mock.ts";
+export type { Spy } from "https://deno.land/std@0.177.0/testing/mock.ts";
```
nit: Update the `mock` above to use the same latest lib version

filmaj replied:
Now that I'm looking to update the std version, I noticed the following:

- On the STD README, under [Releases](https://deno.land/std@0.180.0#releases), it says "To check compatibility of different version of standard library with Deno CLI see [this list](https://raw.githubusercontent.com/denoland/dotland/main/versions.json)."
- [Said list](https://raw.githubusercontent.com/denoland/dotland/main/versions.json) (at the bottom) maps minimum deno version to different std versions

Sooooo, does this mean, when evaluating what version of the stdlib to use, we should take into consideration whatever minimum deno version we want to support for our users? gonna move this convo to a Slack thread...

shapirone reviewed (approved):
LGTM, two small nits!
